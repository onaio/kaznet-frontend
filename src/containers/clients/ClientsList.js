import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Link } from "react-router-dom";
import Moment from "react-moment";
import qs from "qs";
import "../LoadListAnimation.css";
import * as clientActions from "../../store/clients/actions";
import * as clientSelectors from "../../store/clients/reducer";
import * as globalSelectors from "../../store/global/reducer";
import * as globalActions from "../../store/global/actions";
import * as constants from "../../constants.js";

import ListView from "../../components/ListView";
import NotFound from "../../components/NotFound";
import ElementMap from "../ElementMap";

export class ClientsList extends Component {
  async componentDidMount() {
    this.props.showListTitle();
    this.props.changePageTitle("Clients");
    this.props.changePageTitleButton("+ Add Client");
    this.props.changePageTarget("/clients/new");
    let { search } = qs.parse(this.props.location.search.slice(1));
    const { page } = qs.parse(this.props.location.search.slice(1));

    if (search === undefined) {
      search = "";
    }
    this.props.searchVal(search);
    let pageNumber = Number(page);

    if (isNaN(pageNumber)) {
      pageNumber = 1;
    }
    await this.props.fetchClients(
      `${constants.API_ENDPOINT}/clients/?search=${search}&page=${pageNumber}`
    );
    this.props.changePageNumber(pageNumber);
  }

  componentDidUpdate(prevProps) {
    let { search } = qs.parse(this.props.location.search.slice(1));
    if (search === undefined) {
      search = "";
    }
    const { page } = qs.parse(this.props.location.search.slice(1));
    if (Number(page) !== this.props.currentPage && !isNaN(page)) {
      const pageNumber = Number(page);
      this.props.fetchClients(
        `${constants.API_ENDPOINT}/clients/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    }
  }

  render() {
    if (this.props.searchParam !== "" && this.props.rowsIdArray.length === 0) {
      return <NotFound searchVal={this.props.searchParam} />;
    }
    if (this.props.rowsIdArray.length <= 0) return this.renderLoading();
    return (
      <div className="ClientsList">
        <ListView
          renderHeaders={this.renderHeaders}
          rowsIdArray={this.props.rowsIdArray}
          rowsById={this.props.rowsById}
          renderRow={this.renderRow}
          endpoint={"clients"}
          pageLinks={this.props.pageLinks}
          totalPages={this.props.totalPages}
          currentPage={this.props.currentPage}
          firstPage={this.props.firstPage}
          lastPage={this.props.lastPage}
          searchVal={this.props.searchParam}
        />
      </div>
    );
  }

  renderLoading() {
    return (
      <center>
        <div className="lds-ripple">
          <div />
          <div />
        </div>
      </center>
    );
  }

  renderHeaders() {
    const headerItems = ["Name", "Created"];
    return <ElementMap items={headerItems} HTMLTag="th" />;
  }

  renderRow(row) {
    const rowItems = [
      <Link to={`/clients/edit/${row.id}`} key="link_to">
        {row.attributes.name}
      </Link>,
      <Moment key={row.id} format="DD-MM-YYYY">
        {row.attributes.created}
      </Moment>
    ];
    return <ElementMap items={rowItems} HTMLTag="td" />;
  }
}

function mapStateToProps(state) {
  return {
    rowsById: clientSelectors.getClientsById(state),
    rowsIdArray: clientSelectors.getClientsIdArray(state),
    totalPages: clientSelectors.getTotalPages(state),
    currentPage: clientSelectors.getCurrentPage(state),
    pageLinks: clientSelectors.getPageLinks(state),
    firstPage: clientSelectors.getFirstPage(state),
    lastPage: clientSelectors.getLastPage(state),
    searchParam: globalSelectors.getSearchValue(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchClients: clientActions.fetchClients,
      changePageNumber: clientActions.changePageNumber,
      changePageTitle: globalActions.changePageTitle,
      changePageTitleButton: globalActions.changePageTitleButton,
      changePageTarget: globalActions.changePageTarget,
      showListTitle: globalActions.toggleDetailTitleOff,
      searchVal: globalActions.getSearchVal
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ClientsList);
