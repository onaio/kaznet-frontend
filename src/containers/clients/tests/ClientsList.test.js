import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { ClientsList } from "../ClientsList";
import * as fixtures from "../../../store/clients/tests/fixtures";
const history = createBrowserHistory();

describe("containers/clients/ClientsList", () => {
  it("renders without crashing", () => {
    shallow(
      <ClientsList
        fetchClients={function() {}}
        changePageTitle={function() {}}
        changePageTitleButton={function() {}}
        changePageTarget={function() {}}
        changePageNumber={function() {}}
        showListTitle={function() {}}
      />
    );
  });

  it("renders client list correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <ClientsList
          fetchClients={function() {}}
          changePageTitle={function() {}}
          changePageTitleButton={function() {}}
          changePageTarget={function() {}}
          changePageNumber={function() {}}
          showListTitle={function() {}}
          rowsById={fixtures.clientsById}
          endpoint={"clients"}
          rowsIdArray={fixtures.clientsIdArray}
          pageLinks={fixtures.clientData.links}
          totalPages={fixtures.totalPages}
          currentPage={fixtures.clientData.meta.pagination.page}
        />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
