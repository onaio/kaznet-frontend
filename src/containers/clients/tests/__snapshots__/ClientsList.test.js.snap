// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/clients/ClientsList renders client list correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ClientsList
    changePageNumber={[Function]}
    changePageTarget={[Function]}
    changePageTitle={[Function]}
    changePageTitleButton={[Function]}
    currentPage={1}
    endpoint="clients"
    fetchClients={[Function]}
    firstPage={1}
    lastPage={2}
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/clients/?search=&page=1",
        "last": "http://localhost:8000/api/v1/clients/?search=&page=2",
        "next": "http://localhost:8000/api/v1/clients/?search=&page=2",
        "prev": null,
      }
    }
    rowsById={
      Object {
        "1": Object {
          "attributes": Object {
            "created": "2018-06-21T13:51:54.011373+03:00",
            "name": "Sol",
          },
          "id": "1",
          "type": "Client",
        },
      }
    }
    rowsIdArray={
      Array [
        "1",
      ]
    }
    searchParam=""
    searchVal={[Function]}
    showListTitle={[Function]}
    totalPages={2}
  >
    <div
      className="ClientsList"
    >
      <ListView
        currentPage={1}
        endpoint="clients"
        firstPage={1}
        lastPage={2}
        pageLinks={
          Object {
            "first": "http://localhost:8000/api/v1/clients/?search=&page=1",
            "last": "http://localhost:8000/api/v1/clients/?search=&page=2",
            "next": "http://localhost:8000/api/v1/clients/?search=&page=2",
            "prev": null,
          }
        }
        renderHeaders={[Function]}
        renderRow={[Function]}
        rowsById={
          Object {
            "1": Object {
              "attributes": Object {
                "created": "2018-06-21T13:51:54.011373+03:00",
                "name": "Sol",
              },
              "id": "1",
              "type": "Client",
            },
          }
        }
        rowsIdArray={
          Array [
            "1",
          ]
        }
        searchVal=""
        totalPages={2}
      >
        <div>
          <Table
            bordered={true}
            className="kaznet-table"
            responsiveTag="div"
            tag="table"
          >
            <table
              className="kaznet-table table table-bordered"
            >
              <thead>
                <tr>
                  <ElementMap
                    HTMLTag="th"
                    items={
                      Array [
                        "Name",
                        "Created",
                      ]
                    }
                  >
                    <th
                      key="0"
                    >
                      Name
                    </th>
                    <th
                      key="1"
                    >
                      Created
                    </th>
                  </ElementMap>
                </tr>
              </thead>
              <tbody>
                <tr
                  key="1"
                >
                  <ElementMap
                    HTMLTag="td"
                    items={
                      Array [
                        <Link
                          replace={false}
                          to="/clients/edit/1"
                      >
                          Sol
                      </Link>,
                        <t
                          ago={false}
                          calendar={false}
                          decimal={false}
                          element="time"
                          filter={[Function]}
                          format="DD-MM-YYYY"
                          fromNow={false}
                          interval={60000}
                          onChange={[Function]}
                          toNow={false}
                          unit={null}
                          unix={false}
                          utc={false}
                      >
                          2018-06-21T13:51:54.011373+03:00
                      </t>,
                      ]
                    }
                  >
                    <td
                      key="0"
                    >
                      <Link
                        key="link_to"
                        replace={false}
                        to="/clients/edit/1"
                      >
                        <a
                          href="/clients/edit/1"
                          onClick={[Function]}
                        >
                          Sol
                        </a>
                      </Link>
                    </td>
                    <td
                      key="1"
                    >
                      <t
                        ago={false}
                        calendar={false}
                        decimal={false}
                        element="time"
                        filter={[Function]}
                        format="DD-MM-YYYY"
                        fromNow={false}
                        interval={60000}
                        key="1"
                        onChange={[Function]}
                        toNow={false}
                        unit={null}
                        unix={false}
                        utc={false}
                      >
                        <time
                          dateTime={"2018-06-21T10:51:54.011Z"}
                        >
                          21-06-2018
                        </time>
                      </t>
                    </td>
                  </ElementMap>
                </tr>
              </tbody>
            </table>
          </Table>
          <Pagination
            aria-label="Page navigation example"
            listTag="ul"
            tag="nav"
          >
            <nav
              aria-label="Page navigation example"
              className=""
            >
              <ul
                className="pagination"
              >
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="First"
                      className="page-link"
                      replace={false}
                      to="/clients/?search=&status=&page=1"
                    >
                      <a
                        aria-label="First"
                        className="page-link"
                        href="/clients/?search=&status=&page=1"
                        onClick={[Function]}
                      >
                        First
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Previous"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Previous"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        «
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Next"
                      className="page-link"
                      replace={false}
                      to="/clients/?search=&status=&page=2"
                    >
                      <a
                        aria-label="Next"
                        className="page-link"
                        href="/clients/?search=&status=&page=2"
                        onClick={[Function]}
                      >
                        »
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Last"
                      className="page-link"
                      replace={false}
                      to="/clients/?search=&status=&page=2"
                    >
                      <a
                        aria-label="Last"
                        className="page-link"
                        href="/clients/?search=&status=&page=2"
                        onClick={[Function]}
                      >
                        Last
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
              </ul>
            </nav>
          </Pagination>
        </div>
      </ListView>
    </div>
  </ClientsList>
</Router>
`;
